### PR TITLE
Backpropagation

### PR DESCRIPTION
**Pull request by beba1**
Backpropagation implementation

### CODE REVIEW

File: src/Core/backpropagation/CMakeLIsts.txt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+add_library(backpropagation BackupBase.cpp BasicBackup.cpp BackupNegamax.cpp)
```
Comment (Operdies) [inline src/Core/backpropagation/CMakeLIsts.txt]: The "I" is capitalized :)
Rename to CMakeLists.txt

File: src/Core/backpropagation/include/BackupNegamax.h [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#include <BackupBase.h>
+
+class BackupNegmax : BackupBase {
+  public:
+    BackupNegmax();
+    ~BackupNegmax();
+    virtual void backup(SearchNode &node, Reward defaultPolicyScore, double currentPlayer);
+
+  private:
+    double currentPlayer;
```
Comment (Operdies) [inline src/Core/backpropagation/include/BackupNegamax.h]: Kind of pedantic, but shouldn't this be an int?

File: src/Core/DefaultPolicy/DefaultPolicyBase.cpp [hunk 3]
```diff
@@ -1,6 +1,5 @@
 #include <DefaultPolicyBase.h>
 
-DefaultPolicyBase::DefaultPolicyBase(boost::function<States(State<boost::any>&)>& getValidChildStates,
-                                     boost::function<Reward(State<boost::any>&)>& evaluateTerminalState):
-                                     getValidChildStates(getValidChildStates),
-                                     evaluateTerminalState(evaluateTerminalState){};
\ No newline at end of file
+DefaultPolicyBase::DefaultPolicyBase(boost::function<States(State<boost::any> &)> &getValidChildStates,
+                                     boost::function<Reward(State<boost::any> &)> &evaluateTerminalState)
+    : getValidChildStates(getValidChildStates), evaluateTerminalState(evaluateTerminalState){};
```
[the remark this answers is not in the document]
Comment (beba1) [reply src/Core/DefaultPolicy/DefaultPolicyBase.cpp]: Is it really the problem? cause according to this: https://stackoverflow.com/questions/6185020/initializing-member-variables-using-the-same-name-for-constructor-arguments-as-f/6185043#6185043 the compiler knows which is what, and it does initialize class members with arguments?

[the remark this answers is not in the document]
Comment (Operdies) [reply src/Core/DefaultPolicy/DefaultPolicyBase.cpp]: I was sure I had this issue before myself, but if it's in the standard it should be fine. Still think it's weird that scoping rules are different inside and outside of the parentheses, but whatever

File: src/Core/backpropagation/BackupBase.cpp [hunk 4]
```diff
@@ -0,0 +1,12 @@
+#include <BackupBase.h>
+
+BackupBase::BackupBase(){};
+
+// getAncestores - stores simulated node and all the nodes visited in one rollout in one list
```
[the remark this answers is not in the document]
Comment (beba1) [reply src/Core/backpropagation/BackupBase.cpp]: Not sure what you meant by "type in comment" ?

File: src/Core/backpropagation/BasicBackup.cpp [hunk 5]
```diff
@@ -0,0 +1,13 @@
+#include <BasicBackup.h>
+
+BasicBackup::BasicBackup(){};
+
+void BasicBackup::backup(SearchNode &node, Reward defaultPolicyScore) {
+    std::vector<SearchNode *> ancestorsList = {};
+    getAncestors(&node, ancestorsList);
```
[the remark this answers is not in the document]
Comment (Operdies) [reply src/Core/backpropagation/BasicBackup.cpp]: you may be right, I'm testing it in a minute to make sure

File: src/Core/DefaultPolicy/RandomDefaultPolicy.cpp [hunk 6]
```diff
@@ -1,17 +1,17 @@
...
-                                         boost::function<Reward(State<boost::any>&)>& evaluateTerminalState):
-                                         DefaultPolicyBase(getValidChildStates,evaluateTerminalState){};
+RandomDefaultPolicy::RandomDefaultPolicy(boost::function<States(State<boost::any> &)> &getValidChildStates,
+                                         boost::function<Reward(State<boost::any> &)> &evaluateTerminalState)
+    : DefaultPolicyBase(getValidChildStates, evaluateTerminalState){};
 
-Reward RandomDefaultPolicy::defaultPolicy(State<boost::any> state){
+Reward RandomDefaultPolicy::defaultPolicy(State<boost::any> state) {
     States validStates = this->getValidChildStates(state);
     int i;
```
[the remark this answers is not in the document]
Comment (beba1) [reply src/Core/DefaultPolicy/RandomDefaultPolicy.cpp]: Is RandomSampling more descriptive?

File: src/Core/DefaultPolicy/TicTacToeDefaultPolicy.cpp [hunk 7]
```diff
@@ -14,7 +14,7 @@ Reward TicTacToeDefaultPolicy::defaultPolicy(State<boost::any> state) {
 
     while (childStates.size()) {
 
-        // Check if any of the valid states are terminal (have no valid child states)
+        // Checking if any of the valid states are terminal (have no valid child states)
         for (int i = 0; i < childStates.size(); i++) {
```
[the remark this answers is not in the document]
Comment (beba1) [reply src/Core/DefaultPolicy/TicTacToeDefaultPolicy.cpp]: Regarding usage of reference types for states:
I actually ran onto problems when I wanted to reference the states from environment's getValidChildStates method (same thing happens for both pointers and reference(
`//States *validChildStates = &(this->getValidChildStates(state));`

as supposedly it is not legal to reference these temporary instances:
https://stackoverflow.com/questions/16481490/error-taking-address-of-temporary-fpermissive

[the remark this answers is not in the document]
Comment (Operdies) [reply src/Core/DefaultPolicy/TicTacToeDefaultPolicy.cpp]: My main point is that a foreach loop is more readable and reliable when an explicit for-loop isn't required

[the remark this answers is not in the document]
Comment (Operdies) [reply src/Core/DefaultPolicy/TicTacToeDefaultPolicy.cpp]: And I didn't mean to make validChildStates a reference type, only tempState.
it is possible that it is implicity a reference when using State<boost::any>, but if this is the case it is not clear, which is why I suggested iterating with
```
        for (auto& childState : childStates){
            tempState = childState;
```

File: .vscode/launch.json [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+                    "description": "Enable pretty-printing for gdb",
+                    "text": "-enable-pretty-printing",
+                    "ignoreFailures": true
+                }
+            ],
+            "miDebuggerPath": "/usr/bin/gdb"
+        }
+    ]
+}
+
```
Comment (Hoeiriis) [inline .vscode/launch.json]: Add to gitignore and remove

File: .vscode/settings.json [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+                "settings": {
+                    "foreground": "#0f0"
+                }
+            }
+        ]
+    },
+    "gtest-adapter.debugConfig": [
+        "(gdb) Launch"
+    ]
+}
```
Comment (Hoeiriis) [inline .vscode/settings.json]: Add to gitignore and remove

File: src/Core/backpropagation/include/BackupNegamax.h [hunk 10]
```diff
@@ -0,0 +1,11 @@
+#include <BackupBase.h>
+
+class BackupNegmax : BackupBase {
```
Comment (Hoeiriis) [inline src/Core/backpropagation/include/BackupNegamax.h]: Typo in class name, when comparing to the filename: BackupNegamax/BackupNegmax

File: src/Core/backpropagation/include/BackupNegamax.h [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+  public:
+    BackupNegmax();
+    using BackupBase::backup;
```
Comment (Hoeiriis) [inline src/Core/backpropagation/include/BackupNegamax.h]: Is this line necessary for anything?

Comment (beba1) [reply src/Core/backpropagation/include/BackupNegamax.h]: Sooo it is about that name hiding thing, as mentioned [here](https://stackoverflow.com/questions/16835897/overloading-base-class-method-in-derived-class). Problem was that I had to overload the backup function defined in backupBase, and without this it didn't want to compile.

**Pull request merged**